modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/DefaultIRTreeToASMBytesPhase.java

```diff
@@ -584,8 +593,22 @@ public void visitForEachSubIterableLoop(ForEachSubIterableNode irForEachSubItera
         methodWriter.ifZCmp(MethodWriter.EQ, end);
 
         methodWriter.visitVarInsn(iterator.getAsmType().getOpcode(Opcodes.ILOAD), iterator.getSlot());
-        methodWriter.invokeInterface(ITERATOR_TYPE, ITERATOR_NEXT);
-        methodWriter.writeCast(irForEachSubIterableNode.getDecorationValue(IRDCast.class));
+        if (painlessMethod != null || variable.getType().isPrimitive() == false) {
+            methodWriter.invokeInterface(ITERATOR_TYPE, ITERATOR_NEXT);
+            methodWriter.writeCast(irForEachSubIterableNode.getDecorationValue(IRDCast.class));
+        } else {
+            switch (variable.getAsmType().getSort()) {
+                case Type.BOOLEAN -> methodWriter.invokeInterface(VALUE_ITERATOR_TYPE, VALUE_ITERATOR_NEXT_BOOLEAN);
+                case Type.BYTE -> methodWriter.invokeInterface(VALUE_ITERATOR_TYPE, VALUE_ITERATOR_NEXT_BYTE);
```

Comment:
Ok I think I see why ValueIterator is necessary here.  You don't know the underlying def type so you defer that to the iterator implmentation

I was thinking that if the variable is a primitive and the iterator is one of the three, say, primitive iterators, Double, Int, Long, then one could just call those.

However, since we don't know the underlying type, deferring to the code in [`Def.java`](https://github.com/elastic/elasticsearch/pull/92025/files#diff-f1a984f25f3f74d578c91f468085bf1367112e719e81fc8a266e27278f8d9473R745) to perform the cast makes sense.

Reply:
Correct - the actual implementation of the `next*` method depends on what the def actually is. Even if we were to use primitive iterators defined in the JCL, we would still need implementations of those interfaces to cast/convert to the correct type to store in the iteration variable.